From c875b0ec22387f1081224016e7b033fb92d57722 Mon Sep 17 00:00:00 2001
From: Takeshi Saito <takeshi.saito.xv@renesas.com>
Date: Wed, 20 Mar 2019 18:18:10 +0900
Subject: [PATCH 740/909] mmc: renesas_sdhi: max_segs_on_iommu applies only
 no-sdio port

commit caabf4e7396a2b9bef48f180d6dd3644b243c8d1 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

In Gen3 IOMMU, SG entries page size expects a multiple of PAGE_SIZE.
However, SG entries are allocated in small segment in SDIO.
max_segs_on_iommu applies only "no-sdio" port in DeviceTree.

In Atheros SDIO WiFi, the SDIO firmware crashes during the execution
of iperf3 command.

iperf3 -c <IP address>

[   53.457669] ath6kl: firmware crashed
[   53.473327] ath6kl: crash dump:
:

Signed-off-by: Takeshi Saito <takeshi.saito.xv@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mmc/host/renesas_sdhi_core.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/mmc/host/renesas_sdhi_core.c b/drivers/mmc/host/renesas_sdhi_core.c
index 345232bff5b0..0b3d94fe1795 100644
--- a/drivers/mmc/host/renesas_sdhi_core.c
+++ b/drivers/mmc/host/renesas_sdhi_core.c
@@ -853,7 +853,9 @@ int renesas_sdhi_probe(struct platform_device *pdev,
 		mmc_data->capabilities2 |= of_data->capabilities2;
 		mmc_data->dma_rx_offset = of_data->dma_rx_offset;
 		mmc_data->max_blk_count = of_data->max_blk_count;
-		if (pdev->dev.iommu_group)
+		/* IOMMU multiple segments applies only No-SDIO port */
+		if (pdev->dev.iommu_group &&
+		    host->mmc->caps2 & MMC_CAP2_NO_SDIO)
 			mmc_data->max_segs = of_data->max_segs_on_iommu;
 		else
 			mmc_data->max_segs = of_data->max_segs;
-- 
2.17.1

